src/conn/conn_tiered.c

```diff
@@ -120,23 +120,158 @@ __tier_storage_remove_local(WT_SESSION_IMPL *session, const char *uri, bool forc
+ */
+int
+__wt_tier_do_flush(
+  WT_SESSION_IMPL *session, WT_TIERED *tiered, const char *local_uri, const char *obj_uri)
+{
+    WT_DECL_RET;
+    WT_FILE_SYSTEM *bucket_fs;
+    WT_STORAGE_SOURCE *storage_source;
+    const char *local_name, *obj_name;
+
+    storage_source = tiered->bstorage->storage_source;
+    bucket_fs = tiered->bstorage->file_system;
+
+    local_name = local_uri;
+    WT_PREFIX_SKIP_REQUIRED(session, local_name, "file:");
+    obj_name = obj_uri;
+    WT_PREFIX_SKIP_REQUIRED(session, obj_name, "object:");
+
+    /* This call make take a while, and may fail due to network timeout. */
+    WT_RET(storage_source->ss_flush(
+      storage_source, &session->iface, bucket_fs, local_name, obj_name, NULL));
+
+#if 1
+    WT_WITH_CHECKPOINT_LOCK(session,
+      WT_WITH_SCHEMA_LOCK(session, ret = __tier_flush_meta(session, tiered, local_uri, obj_uri)));
+    WT_RET(ret);
+#else
+    /*
+     * TODO: This code is used if called from flush_tier code directly instead of internal thread.
+     * When removing this half of the ifdef, also make this function static and remove __wt prefix.
      */
-    WT_RET(__tier_storage_remove_local(session, NULL, 0));
+    WT_RET(__tier_flush_meta(session, tiered, local_uri, obj_uri));
+#endif
+
+    /*
+     * We may need a way to restart flushes for those not completed (after a crash), or failed (due
+     * to previous network outage).
+     */
+    WT_RET(storage_source->ss_flush_finish(
+      storage_source, &session->iface, bucket_fs, local_name, obj_name, NULL));
     return (0);
 }
 
+/*
+ * __wt_tier_flush --
+ *     Given an ID generate the URI names and call the flush code.
+ */
+int
+__wt_tier_flush(WT_SESSION_IMPL *session, WT_TIERED *tiered, uint64_t id)
+{
+    WT_DECL_RET;
+    const char *local_uri, *obj_uri;
+
+    local_uri = obj_uri = NULL;
+    WT_ERR(__wt_tiered_name(session, &tiered->iface, id, WT_TIERED_NAME_LOCAL, &local_uri));
+    WT_ERR(__wt_tiered_name(session, &tiered->iface, id, WT_TIERED_NAME_OBJECT, &obj_uri));
+    WT_ERR(__wt_tier_do_flush(session, tiered, local_uri, obj_uri));
+
+err:
+    __wt_free(session, local_uri);
+    __wt_free(session, obj_uri);
+    return (ret);
+}
+
+/*
+ * __tier_storage_copy --
+ *     Perform one iteration of copying newly flushed objects to the shared storage.
+ */
+static int
+__tier_storage_copy(WT_SESSION_IMPL *session)
+{
+    WT_DECL_RET;
+    WT_TIERED_WORK_UNIT *entry;
+
+    entry = NULL;
+    for (;;) {
+        /*
+         * We probably need some kind of flush generation so that we don't process flush items for
+         * tables that are added during an in-progress flush_tier. This thread could run due to a
+         * condition timeout rather than a signal. Checking that generation number would be part of
+         * calling __wt_tiered_get_flush so that we don't pull it off the queue until we're sure we
+         * want to process it.
```

Comment:
Is flush_tier exclusive, like checkpoint?  If so, I don't see how we can be adding more entries until this one is complete.